### PR TITLE
Mailbox 297 Optimize ranges flags updates

### PR DESCRIPTION
**Pull request by chibenwa**
[Apache issue](https://issues.apache.org/jira/browse/MAILBOX-297)

This PR proposes a x42 time improvement on flags updates for 5000 messages.

These results were obtained on master using https://github.com/linagora/gatling-imap/pull/24 :

screenshot from 2017-05-23 10-27-26

Then I avoid reading messages (lead to this improvement):

screenshot from 2017-05-23 10-27-28

Then reuse modseq value, where possible, implementing staged updates:

screenshot from 2017-05-23 10-27-30

And finally, parallelised stage operations, by batches of 20 messages.

screenshot from 2017-05-23 10-27-31

### CODE REVIEW

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 1]
```diff
@@ -288,41 +290,49 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+                    result.getSucceeded().stream()
+                        .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)))
+                    .completableFuture()
+                    .thenApply(any -> result)))
+            .map(CompletableFuture::join));
+    }
 
-        return result;
+    private FlagsUpdateStageResult reduceResult(Stream<FlagsUpdateStageResult> stream) {
+        return stream.reduce(FlagsUpdateStageResult::merge).orElse(FlagsUpdateStageResult.none());
```
Comment (quynhn) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: one more bread on "orElse"

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 2]
```diff
@@ -416,9 +414,6 @@ private boolean updateFlags(MailboxMessage message, long oldModSeq) {
             .orElseThrow(MailboxDeleteDuringUpdateException::new);
         return tryMessageFlagsUpdate(flagUpdateCalculator,
                 mailbox,
-                messageDAO.retrieveMessages(ImmutableList.of(composedMessageIdWithMetaData), FetchType.Metadata, Optional.empty()).join()
-                    .findFirst()
-                    .map(pair -> pair.getLeft().toMailboxMessage(ImmutableList.of()))
-                    .orElseThrow(() -> new MessageDeletedDuringFlagsUpdateException(cassandraId, (CassandraMessageId) messageId)));
+                composedMessageIdWithMetaData);
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: isn't it a problem to not reload metadata from db on retry ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 3]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
-        return messageIdDAO.retrieveMessages(mailboxId, set)
-                .join()
-                .flatMap(message -> updateFlagsOnMessage(mailbox, flagUpdateCalculator, message))
-                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                    .thenApply(voidValue -> updatedFlags))
-                .map(CompletableFuture::join)
-                .collect(Collectors.toList()) // This collect is here as we need to consume all the stream before returning result
-                .iterator();
+
+        return runUpdate(mailboxId, set, flagUpdateCalculator).iterator();
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: what is the point of having this method only delegate to another one ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 4]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+    }
+
+    private List<UpdatedFlags> runUpdate(CassandraId mailboxId, MessageRange set, FlagsUpdateCalculator flagsUpdateCalculator) throws MailboxException {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I suggest messageIds

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 5]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, set).join();
+
+        FlagsUpdateStageResult globalResult = runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator);
+
+        int retryCount = 0;
+
+        while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: you mix different level of abstraction : you should extract this loop in a method for this method to be consistent, don't you think ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 6]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+            retryCount++;
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId,
+                FluentFutureStream.of(
+                    globalResult.getFailed().stream()
+                        .map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
+                    .flatMap(OptionalConverter::toStream)
+                    .completableFuture().join(),
+                flagsUpdateCalculator);
+
+            globalResult = globalResult.keepSuccess().merge(stageResult);
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: variable re-allocation ! are you crazy ? (:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 7]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, set).join();
+
+        FlagsUpdateStageResult globalResult = runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator);
+
+        int retryCount = 0;
+
+        while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
+            retryCount++;
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId,
+                FluentFutureStream.of(
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: If I understand your probably-too-smart algorithm, you retry the update on failed flags only, isn't ? I suggest to find a way to write it so that everybody can understand in less than 2 minutes (I needed more than that and not sure I'm right)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 8]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+                    globalResult.getFailed().stream()
+                        .map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
+                    .flatMap(OptionalConverter::toStream)
+                    .completableFuture().join(),
+                flagsUpdateCalculator);
+
+            globalResult = globalResult.keepSuccess().merge(stageResult);
+        }
+
+        LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: will it display IMAP UIDs or store ids ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 9]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+
+        FlagsUpdateStageResult result = toBeUpdated
+            .map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator,
+                newModSeq,
+                oldMetadata))
+            .reduce(FlagsUpdateStageResult::merge)
+            .orElse(none());
+
+        result.getSucceeded().stream()
+            .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: nothing wrong can happen here ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 10]
```diff
@@ -283,14 +278,51 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+
+            globalResult = globalResult.keepSuccess().merge(stageResult);
+        }
+
+        LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
+
+        return globalResult.getSucceeded();
+    }
+
+    private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Why this method has this name ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 11]
```diff
@@ -388,32 +409,49 @@ private boolean updateFlags(ComposedMessageIdWithMetaData oldMetadata, Flags new
...
+                    .addAll(this.failed)
+                    .addAll(other.failed)
+                    .build(),
+                ImmutableList.<UpdatedFlags>builder()
+                    .addAll(this.succeeded)
+                    .addAll(other.succeeded)
+                    .build());
+        }
+
+        public FlagsUpdateStageResult keepSuccess() {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: keepSucceeded ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 12]
```diff
@@ -66,6 +67,7 @@
         .build();
     public static final int EXPUNGE_BATCH_SIZE = 100;
     public static final Logger LOGGER = LoggerFactory.getLogger(CassandraMessageMapper.class);
+    public static final int UPDATE_FLAGS_BATCH_SIZE = 20;
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: could you group it with EXPUNGE_BATCH_SIZE ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 13]
```diff
@@ -288,41 +290,49 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+                .flatMap(OptionalConverter::toStream)
+                .join();
+
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId, idsFailed, flagsUpdateCalculator);
 
             globalResult = globalResult.keepSuccess().merge(stageResult);
         }
 
-        LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
+        if (!globalResult.getFailed().isEmpty()) {
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: it looks unrelated to this commit, isn't it ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 14]
```diff
@@ -288,41 +290,49 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+        return runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator, newModSeq);
+    }
 
-        result.getSucceeded().stream()
-            .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                .thenApply(voidValue -> updatedFlags))
-            .forEach(CompletableFuture::join);
+    private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator, Long newModSeq) {
+        return reduceResult(toBeUpdated.collect(JamesCollectors.chunker(UPDATE_FLAGS_BATCH_SIZE))
+            .values().stream()
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: can't we have a chunker that doesn't require calling `values()` ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 15]
```diff
@@ -288,41 +290,49 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+            .map(uidChunk -> FluentFutureStream.of(
+                uidChunk.stream().map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator, newModSeq, oldMetadata)))
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I find this chunk of code difficult to read because it mixes very technical stuff with what is actually important. Could you try to make the code simpler ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    public static FlagsUpdateStageResult none() {
+        return new FlagsUpdateStageResult(ImmutableList.of(), ImmutableList.of());
+    }
+
+    private final List<MessageUid> failed;
+    private final List<UpdatedFlags> succeeded;
+
+    @VisibleForTesting
+    FlagsUpdateStageResult(List<MessageUid> failed, List<UpdatedFlags> succeeded) {
+        this.failed = failed;
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java]: you should copy the lists or make sure it's ImmutableList

**Comment by rouazana** [comment]
Is there any reason for your APPENDs to be fastest?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 12 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I do believe we can expose these batch size with use of @aduprat  PR (once merged) to allow admin to configure it. I do belive it might depend on hardware you are using...

Yes for grouping.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 5 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Will try, yes...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 14 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Good idea

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 7 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Yes it is, retry is only performed on previously failed messages. I guess method extraction will help for this.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 15 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: That's the problem with // every thing. I should put chunkers everywhere to ensure not overwhelming Cassandra.

I'd like to say metod extraction will help.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 6 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Actually I found **while** to be the easiest way to express this, and recursion was much of a mess....

Thus I kept it like this.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 8 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Stored UIDs, that we can actually use, and read from a sysadmin perspective.

**Comment by chibenwa** [comment]
@rouazana APPEND mean response time is stable (also checked MIN, 50%, 99% have 10% variation, and only MAX show significant changes on last run).

I'd put the fault on statistic instability...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java (hunk 16 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java]: Ok, but we already get this guaranty. (none, success, fail, merge and keepSuccess, which are the only ways to create this object already ensure it).

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 11 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Ok

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 9 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: It's writes, we're efficient at it. 

An analysis shows that TRACE query on the prod that failed actually all are reads.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: It's reloaded on each stage, so it's fresh.

We should also limit READs to the minimum as they are expensive.

So no we should not.

And BTW, it pass concurrency tests, which leads to no problem.

**Comment by rouazana** [comment]
I was looking at req/s which is increasing. Funny how looking at different metrics you can conclude different things :)

**Comment by chibenwa** [comment]
Well it's good news @rouazana ...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: It's bad either, it's a range.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 15 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I extracted so that it's named and thus not obfuscated

**Comment by chibenwa** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 17]
```diff
@@ -278,61 +278,69 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+            .map(uidChunk -> performUpdatesForChunk(mailboxId, flagsUpdateCalculator, newModSeq, uidChunk))
             .map(CompletableFuture::join));
     }
 
+    private CompletableFuture<FlagsUpdateStageResult> performUpdatesForChunk(CassandraId mailboxId, FlagsUpdateCalculator flagsUpdateCalculator, Long newModSeq, List<ComposedMessageIdWithMetaData> uidChunk) {
+        Stream<CompletableFuture<FlagsUpdateStageResult>> updateMetaDataFuture =
+            uidChunk.stream().map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator, newModSeq, oldMetadata));
+
+        return FluentFutureStream.of(updateMetaDataFuture)
+            .completableFuture()
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I think you can use FluentFutureStream#reduce directly

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java [hunk 18]
```diff
@@ -46,8 +46,8 @@ public static FlagsUpdateStageResult none() {
 
     @VisibleForTesting
     FlagsUpdateStageResult(List<MessageUid> failed, List<UpdatedFlags> succeeded) {
-        this.failed = failed;
-        this.succeeded = succeeded;
+        this.failed = ImmutableList.copyOf(failed);
```
Comment (strokyl) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/utils/FlagsUpdateStageResult.java]: I would rather have changed the argument type to ImmutableList, this is private you could change them anytime if you want and it will avoid this copyOf

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 19]
```diff
@@ -283,7 +283,7 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
     @Override
     public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagUpdateCalculator, MessageRange set) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
-        return retrieveMessages(retrieveMessageIds(mailboxId, set), FetchType.Metadata, Optional.empty())
+        return messageIdDAO.retrieveMessages(mailboxId, set)
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: outdated comment. Fixed in fixups.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 9 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Well, if a meteor falls on our servers, as we do not have transaction semantic across tables, this is not something that can be solved easily.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 5 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: What do you think of last version ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 20]
```diff
@@ -281,16 +279,63 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
-                .flatMap(message -> updateFlagsOnMessage(mailbox, flagUpdateCalculator, message))
-                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                    .thenApply(voidValue -> updatedFlags))
-                .map(CompletableFuture::join)
-                .collect(Collectors.toList()) // This collect is here as we need to consume all the stream before returning result
-                .iterator();
+
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, range).join();
+
+        FlagsUpdateStageResult firstResult = runUpdateStage(mailboxId, toBeUpdated, flagUpdateCalculator);
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Ok, I understand : you would not want to create a Result with failed entries in it just for the purpose of avoiding method duplication ?
What if you don't call that failed but remaining ? you could init a FlagsUpdateCommand with remaining == range and strip the *retry methods ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 21]
```diff
@@ -281,16 +279,63 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
+
+        return FluentFutureStream.of(updateMetaDataFuture)
+            .reduce(FlagsUpdateStageResult.none(), FlagsUpdateStageResult::merge)
+            .thenCompose(result -> updateIndexesForUpdatesResult(mailboxId, result));
+    }
+
+    private CompletableFuture<FlagsUpdateStageResult> updateIndexesForUpdatesResult(CassandraId mailboxId, FlagsUpdateStageResult result) {
+        return FluentFutureStream.of(
+            result.getSucceeded().stream()
+                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)))
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: what if update index fail ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 20 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: It's an optimization : On the first run, we don't need to collect UID one by one, we can use a Cassandra range queries.

As on retries, UID might (will) not form a range, we are forced to fetch them one by one.

This internal difference is what prevent us from factorizing this.

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 22]
```diff
@@ -281,16 +279,63 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
...
-                    .thenApply(voidValue -> updatedFlags))
-                .map(CompletableFuture::join)
-                .collect(Collectors.toList()) // This collect is here as we need to consume all the stream before returning result
-                .iterator();
+
+        Stream<ComposedMessageIdWithMetaData> toBeUpdated = messageIdDAO.retrieveMessages(mailboxId, range).join();
+
+        FlagsUpdateStageResult firstResult = runUpdateStage(mailboxId, toBeUpdated, flagUpdateCalculator);
+        FlagsUpdateStageResult finalResult = handleUpdatesStagedRetry(mailboxId, flagUpdateCalculator, firstResult);
+        if (!finalResult.getFailed().isEmpty()) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I would say **containsFailedResults**

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 21 above)
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: 1. They are not expected to fail because we don't use lightweight conditions here

2. The entire update fails as a Runtime exception is thrown. This might lead to index inconsistencies, that we don't repair yet.

But I don't see how it is different from old code behaviour. 

I think table healing should be a topic on its own. I don't think we can tacle the issue here.

About general failure: Would you like me to LOG and Swallow index update fail ? to not abort the entire result?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 22 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: you ask a result if it contains failed results ? I  think a result is a failure or not (or if it is too strong because it partially succeeded, let implement `succeed` instead)

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 20 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: what about building ranges even for failures ?

File: server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java [hunk 23]
```diff
@@ -361,4 +361,31 @@ public void reduceShouldWork() {
                 .join())
             .contains(6L);
     }
+
+    @Test
+    public void reduceShouldEmptyAccumulatorWhenNoValue() {
```
Comment (strokyl) [inline server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java]: reduceShouldEmptyAccumulatorWhenNoValue -> reduceShouldReturnIdentityAccumulatorWhenNoValue

File: server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java [hunk 24]
```diff
@@ -361,4 +361,31 @@ public void reduceShouldWork() {
...
+        long emptyAccumulator = 0L;
```
Comment (strokyl) [inline server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java]: s/emptyAccumulator/identityAccumulator

File: server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java [hunk 25]
```diff
@@ -361,4 +361,31 @@ public void reduceShouldWork() {
...
+            CompletableFutureUtil.reduce(
+                (i, j) -> i + j,
+                CompletableFutureUtil.<Long>allOfArray(),
+                emptyAccumulator)
+                .join())
+            .isEqualTo(emptyAccumulator);
+    }
+
+    @Test
+    public void reduceShouldWorkWithEmptyAccumulator() {
```
Comment (strokyl) [inline server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java]: reduceShouldWorkWithEmptyAccumulator -> reduceShouldWorkWithIdentityAccumulator

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 20 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: what about this question ?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I thought about it, but it brings significant complexity, as we will end up with a recursion tree and not a simple recursion. 

We should also be careful generating the modseq only once for each stage to not loose most of the benefits from this PR.

Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: We agree recursion tree is complex but I'm almost sure we can avoid it, I'm trying to suggest things to remove complexity (doing the same thing twice with different code) not adding complexity.
If you can't find a way, let me know, I'll approve the PR as-is.

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: The issue here is to limit modseq refresh to the minimum required in an understandable way...

Horizontal Traversal of the execution tree needs a temporary map. We also end up having synchronization on the same level of the execution tree (you want to generate the modseq once per level)...

Even designing the algorithm to work on each stage with ranges and not just UIDs is not trivial, and more complicated than the current solution, even if it starts with a good idea...

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Moreover, you want the update to be parralelized by stage and not by range per stage :

```
I update 1 -> 20

If 2, 14 and 18 fail, I want to update those in //
```

If from an execution point of view I separate them by range, this do not get well with the "I want to limit update batch size" statement...

Ie : 

```
I update UID 1 -> 10.000

If 15  25 64 135 47 and UID 5.001 -> 5.125 fails (200 failures) I should be able to execute it in 10 batches, which means overlap some merges....

With handling retries by merges I don't see any complexity reduction... Just from a theorical point of view I introduced to big conceptual challenges, that will come with their implementation complexity...

File: server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java [hunk 26]
```diff
@@ -361,4 +361,31 @@ public void reduceShouldWork() {
                 .join())
             .contains(6L);
     }
+
+    @Test
+    public void reduceShouldIdentityAccumulatorWhenNoValue() {
```
Comment (rouazana) [inline server/container/util-java8/src/test/java/org/apache/james/util/CompletableFutureUtilTest.java]: *ShouldReturn

**Comment by chibenwa** [comment]
Merged.
